src/analysis/calculus/mean_value.lean

```diff
@@ -42,84 +42,256 @@ set_option class.instance_max_depth 120
-implies a bound on the distance of the endpoints images -/
-theorem norm_image_sub_le_of_norm_deriv_le_segment {f : ℝ → F} {C : ℝ}
-  (hf : differentiable_on ℝ f (Icc 0 1))
-  (bound : ∀t ∈ Icc (0:ℝ) 1, ∥deriv_within f (Icc 0 1) t∥ ≤ C) :
-  ∥f 1 - f 0∥ ≤ C :=
+lemma image_le_of_deriv_right_lt_deriv_boundary' {f f' : ℝ → ℝ} {a b : ℝ}
+  (hf : continuous_on f (Icc a b))
+  (hf' : ∀ x ∈ Ico a b, has_deriv_within_at f (f' x) (Ioi x) x)
+  {B B' : ℝ → ℝ} (ha : f a ≤ B a) (hB : continuous_on B (Icc a b))
+  (hB' : ∀ x ∈ Ico a b, has_deriv_within_at B (B' x) (Ioi x) x)
+  (bound : ∀ x ∈ Ico a b, f x = B x → f' x < B' x) :
+  ∀ ⦃x⦄, x ∈ Icc a b → f x ≤ B x :=
+image_le_of_liminf_slope_right_lt_deriv_boundary hf
+  (λ x hx r hr, (hf' x hx).liminf_right_slope_le hr) ha hB hB' bound
+
+lemma image_le_of_deriv_right_lt_deriv_boundary {f f' : ℝ → ℝ} {a b : ℝ}
+  (hf : continuous_on f (Icc a b))
+  (hf' : ∀ x ∈ Ico a b, has_deriv_within_at f (f' x) (Ioi x) x)
+  {B B' : ℝ → ℝ} (ha : f a ≤ B a) (hB : ∀ x, has_deriv_at B (B' x) x)
+  (bound : ∀ x ∈ Ico a b, f x = B x → f' x < B' x) :
+  ∀ ⦃x⦄, x ∈ Icc a b → f x ≤ B x :=
+image_le_of_deriv_right_lt_deriv_boundary' hf hf' ha
+  (λ x hx, (hB x).continuous_at.continuous_within_at)
+  (λ x hx, (hB x).has_deriv_within_at) bound
+
+lemma image_le_of_deriv_right_le_deriv_boundary {f f' : ℝ → ℝ} {a b : ℝ}
+  (hf : continuous_on f (Icc a b))
+  (hf' : ∀ x ∈ Ico a b, has_deriv_within_at f (f' x) (Ioi x) x)
+  {B B' : ℝ → ℝ} (ha : f a ≤ B a) (hB : continuous_on B (Icc a b))
+  (hB' : ∀ x ∈ Ico a b, has_deriv_within_at B (B' x) (Ioi x) x)
+  (bound : ∀ x ∈ Ico a b, f' x ≤ B' x) :
+  ∀ ⦃x⦄, x ∈ Icc a b → f x ≤ B x :=
+image_le_of_liminf_slope_right_le_deriv_boundary hf ha hB hB' $
+assume x hx r hr, (hf' x hx).liminf_right_slope_le (lt_of_le_of_lt (bound x hx) hr)
+
+/-! ### Vector-valued functions `f : ℝ → E` -/
+
+section
+
+variables {f : ℝ → E} {a b : ℝ}
+
+/-- General fencing theorem for continuous functions with an estimate on the derivative.
+Let `f` and `B` be continuous functions on `[a, b]` such that
+
+* `∥f a∥ ≤ B a`;
+* `B` has a right derivative at every point `[a, b)`;
+* for each `x ∈ [a, b)` the right-side limit inferior of the norm of `(f z - f x) / (z - x)`
+  is bounded above by a function `f'` which is strictly less than `B'`.
+
+Then `∥f x∥ ≤ B x` everywhere on `[a, b]`. Most probably you want to use
+`image_norm_le_of_norm_deriv_right_le_deriv_boundary`. -/
+lemma image_norm_le_of_liminf_right_slope_norm_lt_deriv_boundary {E : Type*} [normed_group E]
+  {f : ℝ → E} {f' : ℝ → ℝ} (hf : continuous_on f (Icc a b))
+  -- `hf'` actually says `liminf ∥z - x∥⁻¹ * (∥f z∥ - ∥f x∥) ≤ f' x`
+  (hf' : ∀ x ∈ Ico a b, ∀ r, f' x < r →
+    ∃ᶠ z in nhds_within x (Ioi x), (z - x)⁻¹ * (∥f z∥ - ∥f x∥) < r)
+  {B B' : ℝ → ℝ} (ha : ∥f a∥ ≤ B a) (hB : continuous_on B (Icc a b))
+  (hB' : ∀ x ∈ Ico a b, has_deriv_within_at B (B' x) (Ioi x) x)
+  (bound : ∀ x ∈ Ico a b, ∥f x∥ = B x → f' x < B' x) :
```

Comment:
The docstring does not agree with the statement: you only require `f' < B'` when `||f|| = B`.

Reply:
Sometimes I wish we had per-argument docstrings (transforming, e.g., into `<dl>` in the output HTML).